### PR TITLE
HDDS-5599.  [Ozone-Streaming]drop BufferPool and ChunkBuffer to avoid buffer copying

### PR DESCRIPTION
**Pull request by captainzmc**
## What changes were proposed in this pull request?

In BlockDataStreamOutput, we should write directly to the underlying stream. We should not use BufferPool and ChunkBuffer in BlockDataStreamOutput anymore since we want to avoid buffer copying.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5599

## How was this patch tested?

Existing UT

### CODE REVIEW

**Comment by captainzmc** [comment]
Hi @szetszwo @bshashikant @mukul1987 
I've removed the BufferPool dependency in this PR. The ChunkBuffer is not completely removed due to the checksum, so maybe we can discuss it here.
Also, since BufferPool is removed, Retry and Flush policy will be different, and I think we can discuss their implementations in this PR.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockDataStreamOutput.java [hunk 1]
```diff
@@ -426,16 +294,6 @@ private void watchForCommit(boolean bufferFull) throws IOException {
...
-    final List<ChunkBuffer> byteBufferList;
-    if (!force) {
-      Preconditions.checkNotNull(bufferList);
-      byteBufferList = bufferList;
-      bufferList = null;
-      Preconditions.checkNotNull(byteBufferList);
-    } else {
-      byteBufferList = null;
-    }
-
```
Comment (szetszwo) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockDataStreamOutput.java:438]: Move the try-catch below to flush() and then call flush() here.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/CommitWatcher.java [hunk 2]
```diff
@@ -84,6 +84,13 @@ public CommitWatcher(BufferPool bufferPool, XceiverClientSpi xceiverClient) {
     futureMap = new ConcurrentHashMap<>();
   }
 
+  public CommitWatcher(XceiverClientSpi xceiverClient) {
```
Comment (szetszwo) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/CommitWatcher.java:87]: Let's copy CommitWatcher to a new class, say StreamCommitWatcher?  In StreamCommitWatcher, we don't need bufferPool.  The commitIndex2flushedDataMap should be different -- it should not use ChunkBuffer anymore.

**Review by szetszwo — COMMENTED** [review]
The change looks good.  Some comments inlined.   Thanks.

**Review by szetszwo — COMMENTED** [review]
+1, the change looks good.

There are some test failures.  Could you take a look?

**Comment by captainzmc** [comment]
> +1, the change looks good.
> 
> There are some test failures. Could you take a look?

[CI runs correctly](https://github.com/captainzmc/hadoop-ozone/runs/3420162405) on my personal branch. It seems the error is not caused by this PR.  I will try to re-trigger the CI to make sure this.

**Comment by captainzmc** [comment]
hi @szetszwo, CI has been running correct. Could you help take another look this PR?

**Review by szetszwo — APPROVED** [review]
+1

**Pull request merged**